### PR TITLE
Workflow parsing, added metrics and a json metadatafile

### PR DESCRIPTION
**Pull request by Ambivalentalfa**
Three main things in this PR:
- A package development script `pckg_dev.py` is added to the package to do things like parse the data from the APE in the wild paper, which will not be relevant for the final product (should they be somewhere else)
- Updated the package to have one single meta data file in json format containing information about the tools, their publications and the year of their primary publication. Also added a json file to save the citations for each primary publication. This also includes some testing. It is not entirely functional and there is some problem with the doi to pmid method which I have not yet identified.
- Added more metrics and tests for each of them 
- (also general updates to the package, such as not returning tuples as often, and breaking up larger functions.)

Did not upload any data so that has to be generated from scratch to run the new pipeline

### CODE REVIEW

File: src/wfqc/metrics.py [hunk 1]
```diff
@@ -1,81 +1,313 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (vedran-kasalica) [inline src/wfqc/metrics.py:206]:
```suggestion

```

Comment (vedran-kasalica) [reply src/wfqc/metrics.py:206]: Probably the empty lines can be removed

File: tests/test_network.py [hunk 2]
```diff
@@ -2,35 +2,115 @@
...
+citations = ['CA', 'CB', 'CC', 'CD', 'CE', 'CF', 'CG', 'CH', 'CI', 'CJ', 'CK', 'CL', 'CM', 'CN', 'CO', 'CP', 'CQ']
 
-def test_load_citation_network():
-    G, included_tools = nw.create_citation_network(loadData=True) # TODO: if no data then make data or ask if make data?
-    assert len (included_tools) > 0
+
+edges = [
+
+    # Single citations of tools
+    ('CA', 'TA'), ('CB', 'TB'), ('CC', 'TC'), ('CD', 'TD'),
```
Comment (vedran-kasalica) [inline tests/test_network.py:26]: I would suggest creating a separate Python module named `example_network.py` that contains the example data and then use that data in this and tests/test_metrics.py test modules, for easier maintainability. You can then simplify the examples and call `Graph = igraph.Graph.TupleList(edges, directed=True)` and similar overlapping code directly in `example_network.py`.

File: src/wfqc/pckg_dev.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+import pandas as pd
+import json
+import numpy as np
+import json 
+import random
+from sklearn.model_selection import train_test_split
+
+
+def stratified_split(data, test_size=0.2, random_state=42):
+    rating_bins = [0, 1, 2, 3]
```
Comment (vedran-kasalica) [inline src/wfqc/pckg_dev.py:12]: This variable is not used as such right? Maybe the line is not needed then.

File: src/wfqc/pckg_dev.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+
+    for d in data:
+        if d['ratingAvg'] == 0:
+            d['rating_bin'] = '0'
+        elif d['ratingAvg'] > 0 and d['ratingAvg'] < 1:
+            d['rating_bin'] = '0-1'
+        elif d['ratingAvg'] >= 1 and d['ratingAvg'] < 2:
+            d['rating_bin'] = '1-2'
+        elif d['ratingAvg'] >= 2 and d['ratingAvg'] <= 3:
+            d['rating_bin'] = '2-3'
```
Comment (vedran-kasalica) [inline src/wfqc/pckg_dev.py:22]: If this is to be part of the package, using a method that simply rounds up (or down) the number would be more elegant. As it is now, it is a bit confusing, e.g., `0-1` does not include 0 nor 1, but `2-3` includes both 2 and 3.

File: src/wfqc/pckg_dev.py [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+#     # Optionally, save to a file
+#     with open('ratingsOfusecases.json', 'w') as json_file:
+#         json_file.write(json_data)
+
+#     return data
+
+
+
+
+import pandas as pd
```
Comment (vedran-kasalica) [inline src/wfqc/pckg_dev.py:216]: Imports should always be added at the beginning of the module/file

File: src/wfqc/pckg_dev.py [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+import json
+import numpy as np
+import json 
+import random
+from sklearn.model_selection import train_test_split
+
+
+def stratified_split(data, test_size=0.2, random_state=42):
```
Comment (vedran-kasalica) [inline src/wfqc/pckg_dev.py:11]: Even though this module might not be part of the final package it is a good habit to document methods, it is easier to debug, review and improve later on, even if the documentation is one sentence.

File: src/wfqc/metrics.py [hunk 7]
```diff
@@ -1,81 +1,313 @@
...
-    ----------
-    graph : igraph.Graph
-        Graph generated by igraph
-    workflow, list of tuples with pairs of strings TODO: how do I write this? does it matter I wont have this format later anyways? 
-        List of tuples of strings corresponding to the edges in the workflow. 
 
-    """
-    # List to  collect pairwise scores
-    score_list = [] # TODO: can predefine the list length, does not matter this is temporary?
+def get_node_ids(graph: igraph.Graph, key= "name"):
```
Comment (vedran-kasalica) [inline src/wfqc/metrics.py:6]: It is a good practice to use typing to describe the data types, apart from better documentation and readability of the function it is possible to use type checking libraries such as `pytype` to ensure that the code uses the data types as they are intended, e.g., a `string` is never expected to be an `int`, or a `list` is never expected to be a `dict`. In addition, the methods should include docstrings, describing the method, each of the parameters, and the return value.

```suggestion
def get_node_ids(graph: igraph.Graph, key:str= "name") -> dict:
```

File: src/wfqc/metrics.py [hunk 8]
```diff
@@ -1,81 +1,313 @@
...
 
-    return sum(score_list)/len(score_list), score_list
+    if edge[0] not in graph.vs['name'] or edge[1] not in graph.vs['name']:
+        return None 
+    try:
+        source = edge[0]
+        target = edge[1]
+        weight = graph.es.find(_between=((id_dict[source],), (id_dict[target],)))['weight']
+    except:
+        weight = 0
```
Comment (vedran-kasalica) [inline src/wfqc/metrics.py:28]: Is it a valid logic that in the case of a missing edge the method returns `None`, but in the case of an `Exception` the method returns `0`

File: src/wfqc/metrics.py [hunk 9]
```diff
@@ -1,81 +1,313 @@
...
 
+    return weight
 
 
-def path_metric(G, workflow, included_tools): #TODO: how normalise this?
-    # shortest path == shortest path of the subpaths. Thus:
+import math
 
-    shortest_paths = []
+def simple_sum_metric(graph, workflow):
```
Comment (vedran-kasalica) [inline src/wfqc/metrics.py:35]: Each metric must be documented, as they present the crucial features of the package.

File: src/wfqc/metrics.py [hunk 10]
```diff
@@ -1,81 +1,313 @@
...
-        if source in included_tools and target in included_tools:
-            shortest_paths.append(G.get_shortest_paths(source, to = target))
+    if weights:
+        return sum([w for w in weights if w])
+    else: 
+        return 0 # empty workflow has score 0 
+
+def sum_metric(graph, workflow, normalise = True):
+
+    weights = []
```
Comment (vedran-kasalica) [inline src/wfqc/metrics.py:49]: The naming of some variables could be more descriptive, e.g., `edge_weigths` (this is my personal preference).

File: src/wfqc/network.py [hunk 11]
```diff
@@ -15,46 +17,61 @@
...
+import sys
+sys.path.insert(0, os.path.abspath(os.path.join(os.getcwd(), 'src')))
 
-import wfqc.data
+import wfqc.data 
 
 
-def download_data(outpath, testSize, randomSeed, topicID):
+
+async def download_data(outpath, testSize, randomSeed, topicID):
```
Comment (vedran-kasalica) [inline src/wfqc/network.py:27]: The name of the method is quite generic, documentation and param types would be quite useful here.

File: src/wfqc/network.py [hunk 12]
```diff
@@ -230,40 +201,30 @@ def create_citation_network(topicID="topic_0121", testSize='', randomSeed=42, lo
...
-            with open(nr_citations_path, 'wb') as f:
-                pickle.dump(nr_citations, f)
-
             with open(node_degree_dict_path, 'wb') as f:
                 pickle.dump(node_degree_dict, f)
 
 
     # returns a graph and the pmids of the tools included in the graph (tools connected by cocitations)
-    return G, included_tools 
+    return G
```
Comment (vedran-kasalica) [inline src/wfqc/network.py:229]: I would suggest to rename `G` to something more descriptive, in general single letter variable names are not the best practice (unless we are talking about counters (`i`, `j`), coordinates (`x`,`y`), etc. You can leave `f` (line 224) as it clearly represents a file, but in general it's not a good practice. I see another one later in the code, I will leave the comment there as well.

File: src/wfqc/workflow.py [hunk 13]
```diff
@@ -84,3 +84,113 @@ def radnomise_workflows(included_tools, num_pairs = 3):
...
+    return reconnected_edges
+
+def name_to_pmid(file, name):
+    """ Given open csv and a tool name it returns the pmid for that tool""" # TODO change to json
+    pmid = file.loc[file['name'] == name, 'pmid'].values[0] if len(file.loc[file['name'] == name, 'pmid']) > 0 else None
+    return pmid
+    
+def pmid_edges(csv_filename, tool_edges):
+
+    f = pd.read_csv(csv_filename)
```
Comment (vedran-kasalica) [inline src/wfqc/workflow.py:152]: I suggest changing the `f` variable name to something more descriptive.

File: tests/test_data.py [hunk 14]
```diff
@@ -10,18 +15,33 @@ async def test_get_tool_metadata(shared_datadir):
...
     assert df.loc[df['name'] == 'PEPMatch', 'pmid'].values[0] == str(38110863)
 
-def test_europepmc(shared_datadir):
-    ProteoWizard_pmid = 23051804
-    citations = data.europepmc(ProteoWizard_pmid, page_size=1000)
-    assert len(citations)> 100
+
+@pytest.mark.asyncio
+async def test_europepmc_request(shared_datadir):
+    ProteinProphet_pmid = 14632076 #ProteinProphet has 2949 citations on Jul 12th 2024
```
Comment (vedran-kasalica) [inline tests/test_data.py:21]: This variable is specified but is not being used, right? It is also good to avoid using capitalized variable names in Python. You will see often that only class names are capitalized, and all variables, module names, package names, etc. start with a small letter (exceptions are all capital variables in some cases).

**Review by vedran-kasalica — COMMENTED** [review]
The PR looks good, the module structure fits the purpose.
I left comments inline. Some points of improvement:
- methods should be documented (method description, parameters, and return value)
- method params and return value types should be specified if possible 
- variable names should be descriptive, and single-letter variables should be avoided (except for `i,j,x,y` and sometimes `f` for a file when it's a one-line command)

**Important**:
#16 should be merged before this PR gets merged because #16 introduces a GitHub action that tests the code. Those tests should pass before this PR is merged.

File: src/wfqc/pckg_dev.py (hunk 4 above)
Comment (Ambivalentalfa) [reply src/wfqc/pckg_dev.py:22]: This file is just for collecting methods I've written to process data etc, and that I dont think  should be in the final package

Comment (magnuspalmblad) [reply src/wfqc/pckg_dev.py:22]: Yes, I think it would be a good idea to separate scripts for wrangling data used to develop (evaluate) the metric from the code to calculate the metric itself. Maybe these scripts (to create datasets for evaluating the metric) could even be placed in a separate repository where the datasets are also kept?

Comment (vedran-kasalica) [reply src/wfqc/pckg_dev.py:22]: We could indeed create a repo with the "APE in the Wild" data and a Python script used for reading the data, if needed.

Regarding my original question, is the script valid as it is, `0-1` does not include `0` nor `1`, but `2-3` includes both `2` and `3`. Could this affect the quality of the evaluation?
